keps/sig-api-machinery/2835-prioritized-leader-election/README.md

```diff
@@ -0,0 +1,324 @@
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We propose to add an extension to the existing leader election mechanism in client-go to support
+prioritizing election participants by user-defined priority comparison functions.
+
+## Motivation
+
+Prioritized leader election has a few key use cases:
+
+* I have multiple versions of an application running concurrently in the
+cluster; I want the newest one to become the leader.
+* I have multiple replicas with different resources; I want the largest
+replica to be the leader.
+* I am implementing a cross-region controller (think multicluster
+service discovery) and the geographically closest region to be the
+leader, to avoid something like a US controller acting on a cluster in
+Asia, when I have a controller replica already running in Asia.
+
+
+### Goals
+
+* Provide a simple mechanism to allow users to opt into prioritized leader election.
+
+### Non-Goals
+
+* User prioritized leader election in Kubernete's maintained controllers
```

Comment:
in the sig call, I think we decided that this should probably be a goal and it would benefit cases of upgrades with a weight for latest version.